collector/textfile.go

```diff
@@ -52,31 +65,165 @@ func NewTextFileCollector() (Collector, error) {
 	c := &textFileCollector{
 		path: *textFileDirectory,
 	}
+	return c, nil
+}
 
-	if c.path == "" {
-		// This collector is enabled by default, so do not fail if
-		// the flag is not passed.
-		log.Infof("No directory specified, see --collector.textfile.directory")
-	} else {
-		textFileAddOnce.Do(func() {
-			prometheus.DefaultGatherer = prometheus.Gatherers{
-				prometheus.DefaultGatherer,
-				prometheus.GathererFunc(func() ([]*dto.MetricFamily, error) { return c.parseTextFiles(), nil }),
+func convertMetricFamilies(metricFamilies []*dto.MetricFamily, ch chan<- prometheus.Metric) {
+	var valType prometheus.ValueType
+	var val float64
+
+	for _, mf := range metricFamilies {
+		labelNames := map[string]struct{}{}
+		for _, metric := range mf.Metric {
+			labelPairs := metric.GetLabel()
+			for _, label := range labelPairs {
+				if _, ok := labelNames[label.GetName()]; !ok {
+					labelNames[label.GetName()] = struct{}{}
+				}
 			}
-		})
+		}
+
+		for _, metric := range mf.Metric {
+			labelPairs := metric.GetLabel()
+			var labels []string
+			var labelVals []string
+			for _, label := range labelPairs {
+				labels = append(labels, label.GetName())
+				labelVals = append(labelVals, label.GetValue())
+			}
+
+			for k := range labelNames {
+				present := false
+				for _, label := range labels {
+					if k == label {
+						present = true
+						break
+					}
+				}
+				if present == false {
+					labels = append(labels, k)
+					labelVals = append(labelVals, "")
+				}
+			}
+
+			metricType := mf.GetType()
+			switch metricType {
+			case dto.MetricType_COUNTER:
+				if metric.Counter != nil {
```

Comment:
I think we can probably rely on the fact that the parser returns us well-formed metric families (meaning if the metric type is Counter, there should always be a `Counter` field present, and analogous for the other types). At least then if that isn't true, it means there's a programming error and it'll crash and someone will notice. But just hiding such an error silently here isn't a good idea, we'd then have to surface that error differently somehow. I think I'd just remove the `if` condition guards in all these cases.